### ISSUE TITLE
[PAID] [$500] Chat - When opening the mp4 video that was just sent, console error appears

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#37142]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Found when validating PR :** https://github.com/Expensify/App/pull/37007

**Version Number:** v1.4.43-14
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:**

## Action Performed:
1. Open the app
2. Login with any account
3. Start any new chat
4. Click on plus icon inside the compose box
5. Click on add attachment
6. Click on choose from gallery
7. Select an mp4 video
8. When the preview opens up play the video and manually scrub through the video using the progress bar
9. Send a video
10. Open the video after sending it

## Expected Result:
No any error in console

## Actual Result:
Console error appears

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/fdee00bc-6a19-4b3e-9d8b-56abc3008c35

[Bug6389704_1708701973856!Console_23.02.txt](https://github.com/Expensify/App/files/14387003/Bug6389704_1708701973856.Console_23.02.txt)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015cdee1805b9b7baa</li>
        <li>Upwork Job ID: 1761051577601671168</li>
        <li>Last Price Increase: 2024-02-23</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 0</li>
<li>jeremy-croff | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @strepanier03</details>

**Comment by izarutskaya** [comment]
We think that this bug might be related to #vip-vsb
CC @quinthar

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

When opening the mp4 video that was just sent, console error appears.
Same thing happens when the attachment modal opens and we hover over the video controls.

### What is the root cause of that problem?

The main root comes from the video implementation PR https://github.com/Expensify/App/pull/30829, the latest PR which changed code related to the video functionality being PR https://github.com/Expensify/App/pull/37007.

The root cause of the issue is just as the error is mentioning, and the error is thrown by the following `expo-av` Video component function [here](https://github.com/expo/expo/blob/7415c87573d1df05037e15c4f2e48d1a25ddb8d6/packages/expo-av/src/Video.tsx#L96-L108).

<details>
<summary>📜 More context as to why we get this error</summary>
.
The error message "Cannot complete operation because the Video component has not yet loaded" indicates that the code is trying to interact with the Video component before it is fully loaded and ready to be manipulated. This happens because our `playVideo` function calls `expo-av` -> `setStatusAsync` method immediately after rendering, but the native components haven't been set up yet.

In our `playVideo` function, we are correctly checking if `currentVideoPlayerRef` and `currentVideoPlayerRef.current.setStatusAsync` are defined before calling `getStatusAsync`. However, it appears that the error is thrown from within `_performOperationAndHandleStatusAsync`, because the video ref (`!video`) is not yet available.

Since the video plays normally despite the error, this means that the operation is being triggered prematurely, such as before the video is fully prepared to play.
.

</details>

**TLDR:** Since the video plays normally despite the error, this means that the operation is being triggered prematurely, such as before the video is fully prepared to play.

### What changes do you think we should make in order to solve the problem?

To prevent this error, we need to make sure that any operations on the video are performed after it has loaded. We can do this by waiting for the `onLoad` callback to be fired, which indicates that the video is ready.

Here's how we can modify our current `playVideo` function to check if the video is loaded before trying to play it:
- check `!status.isLoaded` and **return;**

**Comment by pecanoro** [comment]
Ok, since this is just a console error and everything seems to be working just fine, I am going to remove the deploy blocker label.

**Comment by jeremy-croff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We get a console error when we play a video in the attachment preview. And then, play a video in the report.

### What is the root cause of that problem?
The root cause of this bug is that the [PlaybackContext.js](https://github.com/Expensify/App/blob/8bdb753045cf0eee6cf1d342adf8b76040cabac4/src/components/VideoPlayerContexts/PlaybackContext.js#L51) is trying to pause the previous video from the modal because currentlyPlayingURL is changing. However, that video context has been destroyed, failing downstream in expo-av like @ikevin127 linked. Our `PlaybackContext.js` still has the stale video reference in `currentVideoPlayerRef.current ` despite it not being loaded.

### What changes do you think we should make to solve the problem?
We currently do a good job cleaning up refs when we navigate between reports because there's a clean-up hook resetVideoPlayerData. This runs on `[currentReportID] `. This, however, requires the video to exist and doesn't run between attachments and reports, for example. The attachment flow is also unique because the blob was only in memory, and when we close the modal, we can no longer clean it up with `resetVideoPlayerData`.

So, how do we ensure that when the attachment video gets destroyed, its references are cleaned up? 

A reliable lifecycle is:
1. In AttachmentsView, we can make sure to set `currentVideoPlayerRef.current = null;` when it unmounts;  
2. Consequently, all the downstream methods in `PlaybackContext` won't execute, like pausing, on their lifecycles methods because they early return on `currentVideoPlayerRef.current = null`
3. In AttachmentsView, we can also leverage [isVideo] variable for executing this logic

This fixes the bug in OP. However, it will fix 2 more bugs that should be considered in all proposals:
1. when playing an existing video in the report and then uploading an attachment and scrubbing it, it throws a console error because of stale reference in trying to pause the blob( without requiring the play the new video)
2. When uploading a video, scrubbing it, and then hovering over the previous videos, it throws console errors because of `volumeContext` trying to reference the blob 

### What alternative solutions did you explore? (Optional)
1. explored managing and resetting the `currentlyPlayingURL` variable; however, it only seems like a partial solution.
2. I explored cleaning up the refs on PlaybackContext.js unmount; however, it is not a reliable lifecycle and breaks other flows

**Comment by jeremy-croff** [comment]
Tested my logic:

https://github.com/Expensify/App/assets/157416545/b1fbbe36-bb24-42dc-af09-c1d964f4f2de

**Comment by alitoshmatov** [comment]
@jeremy-croff Thank you for your proposal. Your RCA is correct and you have a good detailed solution. I tested it and it works as expected.

We can go with @jeremy-croff 's [proposal](https://github.com/Expensify/App/issues/37142#issuecomment-1963050093)
C+ reviewed 🎀 👀 🎀

**Comment by pecanoro** [comment]
Sounds good, assigning @jeremy-croff!

**Comment by strepanier03** [comment]
@jeremy-croff - I've paid you via Upwork and closed the contract. 

@alitoshmatov - I'll keep an eye open for the checklist so post that as soon as you can and I'll finish payment.

**Comment by alitoshmatov** [comment]
- The PR that introduced the bug has been identified. Link to the PR: Related to this PR: https://github.com/Expensify/App/pull/30829
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/30829/files#r1527477209
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/a
- Determine if we should create a regression test for this bug.
No need, since it is a console error issue, and existing regression testing would catch it again.

**Comment by alitoshmatov** [comment]
@strepanier03 Completed the checklist

**Comment by strepanier03** [comment]
Paid @alitoshmatov and closed the contract. Thanks, everyone!

**Comment by olivegardenmenu-crypto** [comment]
The [Olive Garden menu](https://olivegardenmenuwithprices.com/) offers a wide variety of classic Italian-American dishes, from hearty pastas and savory soups to fresh salads and indulgent desserts, all served in a warm and family-friendly atmosphere. Whether you're craving a traditional favorite like Fettuccine Alfredo or exploring new seasonal specialties, there's something for everyone.
